### PR TITLE
Use `ReEmpty(U0)` as the implicit region bound in typeck

### LINKED ISSUE #74429: Borrowck deduces empty lifetime
I tried this code:

```rust
use std::ops::{DivAssign, SubAssign};

use ndarray::{s, Array2, ArrayView1, Axis, Zip};
use num::{Zero, One};
use rayon::prelude::*;

pub fn gaussian_elimination<T>(mut a: Array2<T>) -> Array2<T>
where
    T: One + Zero + Clone + Send + Sync + DivAssign + SubAssign,
{
    let (n, m) = a.dim();
    assert!(n > 0);
    let mut pivots = vec![];
    let mut pivot = 0;
    for i in 0..n {
        if a[[i, pivot]].is_zero() {
            if let Some((new_pivot, row)) = (pivot..m)
                .into_par_iter()
                .flat_map(|pivot| {
                    let a: ArrayView1<_> = a.slice(s![i.., pivot]);
                    a.axis_iter(Axis(0))
                        .into_par_iter()
                        .enumerate()
                        .flat_map(|(j, x)| {
                            if x.into_scalar().is_zero() {
                                None
                            } else {
                                Some(j)
                            }
                        })
                        .find_any(|_| true)
                        .map(|row| (pivot, row))
                })
                .find_first(|_| true)
            {
                pivot = new_pivot;
                if row > 0 {
                    let mut a = a.slice_mut(s![i..; row, ..]);
                    let mut it = a.axis_iter_mut(Axis(0));
                    let this_row = it.next().expect("this row exists");
                    let that_row = it.next().expect("that row exists");
                    Zip::from(this_row).and(that_row).par_apply(std::mem::swap);
                }
            } else {
                break;
            }
        }
        pivots.push((i, pivot));
        let (mut this_equation, mut rest_equations) = a.slice_mut(s![i.., ..]).split_at(Axis(0), 1);
        let mut this_equation = this_equation.index_axis_mut(Axis(0), 0);
        let scale = this_equation[pivot].clone();
        assert!(!scale.is_zero());
        this_equation.iter_mut().for_each(|a| {
            *a /= scale.clone();
        });

        rest_equations
            .axis_iter_mut(Axis(0))
            .into_par_iter()
            .for_each(|mut eqn| {
                let scale = eqn[pivot].clone();
                if !scale.is_zero() {
                    eqn.iter_mut()
                        .zip(this_equation.iter())
                        .for_each(|(e, e_)| *e -= e_.clone() * scale.clone());
                }
            });
        pivot += 1;
        if pivot >= m {
            break;
        }
    }

    // a is in row echelon form now
    if pivots.len() > 1 {
        for (i, pivot) in pivots[1..].iter().rev() {
            let (mut rest_equations, this_equation) =
                a.slice_mut(s![..=*i, ..]).split_at(Axis(0), *i);
            let this_equation = this_equation.index_axis(Axis(0), 0);
            rest_equations.axis_iter_mut(Axis(0)).for_each(|eqn| {
                if eqn[*pivot].is_zero() {
                    return;
                }
                let scale = eqn[*pivot].clone();
                Zip::from(eqn)
                    .and(this_equation)
                    .par_apply(|e, e_| *e -= e_.clone() * scale.clone());
            });
        }
    }

    a
}
```

I expected to see this happen:
On Rust 1.44.0, this compiles.

Instead, this happened:
```
error[E0311]: the parameter type `T` may not live long enough
  --> src/lib.rs:85:17
   |
85 | /                 Zip::from(eqn)
86 | |                     .and(this_equation)
   | |_______________________________________^
   |
   = note: the parameter type `T` must be valid for the empty lifetime...
   = note: ...so that the reference type `&T` does not outlive the data it points at

error: aborting due to previous error
```

### Meta
`rustc --version --verbose`:
```
rustc 1.46.0-nightly (346aec9b0 2020-07-11)
binary: rustc
commit-hash: 346aec9b02f3c74f3fce97fd6bda24709d220e49
commit-date: 2020-07-11
host: x86_64-unknown-linux-gnu
release: 1.46.0-nightly
LLVM version: 10.0
```

### PR DESCRIPTION
**Pull request by matthewjasper**
Fixes #74429

r? @nikomatsakis

### CODE REVIEW

**Comment by LeSeulArtichaut** [comment]
#74429 is a stable regression, so would this PR also deserve a stable backport? cc @Mark-Simulacrum

**Comment by Mark-Simulacrum** [comment]
Seems reasonable, stable-nominating -- @matthewjasper, does that feel reasonable to you? It looks like this is a relatively small patch so I'd feel relatively comfortable backporting it, but you know better of course :)

**Comment by matthewjasper** [comment]
This should be fine for a stable backport

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -684,7 +684,9 @@ impl Inherited<'a, 'tcx> {
             deferred_generator_interiors: RefCell::new(Vec::new()),
             opaque_types: RefCell::new(Default::default()),
             opaque_types_vars: RefCell::new(Default::default()),
-            implicit_region_bound: None,
+            // Make sure that we can always assume `T: ReEmpty(U0)` for any
+            // type parameter `T`.
+            implicit_region_bound: Some(tcx.lifetimes.re_root_empty),
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs:689]: I suspect this field could be made non-optional if we use empty here

Comment (nikomatsakis) [reply src/librustc_typeck/check/mod.rs:689]: I'm a bit confused as to how this winds up fixing any bug -- reading the code that is using this field, it looks like it tries to make `'empty` an implicit bound:

https://github.com/rust-lang/rust/blob/71384101ea3b030b80f7def80a37f67e148518b0/src/librustc_infer/infer/outlives/verify.rs#L73-L99

maybe `VerifyBound::IsEmpty` can be removed after this change, even?

Comment (matthewjasper) [reply src/librustc_typeck/check/mod.rs]: The issue is that it's possible that we have `T: ReEmpty(Un)` as an implied bound from a closure, which causes us to not generate `VerifyBound::IsEmpty`.

**Comment by Mark-Simulacrum** [comment]
@bors p=1

We are most likely backporting to stable so would be good to get it in nightly as quickly as we can for additional testing.

**Comment by pnkfelix** [comment]
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/31818weeklymeeting2020072354818.html#204802756) at weekly T-compiler triage meeting.

beta-accepted

**Comment by pnkfelix** [comment]
the stable-nomination was [also discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/31818weeklymeeting2020072354818.html#204804101) (separately) at the weekly T-compiler triage meeting.

We are a little nervous that this only landed so recently, and yet we have a (stable) point release expected next week.

Still, this is pretty low risk, we think. We will keep our eye out collectively for suspicious behavior on nightly, and revert (at least from stable) if we see anything fishy.

stable-accepted.

**Pull request merged**